src/general/MainMenu.cs

```diff
@@ -752,6 +788,47 @@ private void WarnAboutNoEnabledMods()
         }
     }
 
+    private float TrackMenuPerformance()
+    {
+        var currentFrameRate = Engine.GetFramesPerSecond();
+
+        // If this is the first tracked frame, do not use the average of the frame delta and 0
+        if (averageFrameRate == 0)
+            return currentFrameRate;
+
+        // Not an exact average by any means, but good enough for this purpose
+        return (averageFrameRate + currentFrameRate) / 2;
+    }
+
+    private void WarnAboutLowPerformance()
+    {
+        if (averageFrameRate < 30 && secondsInMenu >= 16 && !AreAnyMenuPopupsOpen())
+        {
+            GD.Print($"Average frame rate is {averageFrameRate}, prompting to disable 3D backgrounds");
+            lowPerformanceWarning.PopupIfNotDismissed();
+            canShowLowPerformanceWarning = false;
+        }
+    }
+
+    private void OnLowPerformanceDialogConfirmed()
+    {
+        Settings.Instance.Menu3DBackgroundEnabled = new SettingValue<bool>(false);
+        Settings.Instance.ApplyGraphicsSettings();
+        Settings.Instance.Save();
+
+        // This doesn't seem to be called automatically if this setting is changed outside the designated method
```

Comment:
In my earlier comment I talked about the observables breaking on reassigning the option value, this is probably caused by that. And this should be removed once that is fixed.